modules/blocks/chain.js

```diff
@@ -736,37 +739,49 @@ Chain.prototype.deleteLastBlock = function(cb) {
 		return setImmediate(cb, 'Cannot delete genesis block');
 	}
 
-	async.waterfall(
-		[
-			waterCb => {
-				// Delete last block, replace last block with previous block, undo things
-				__private.popLastBlock(lastBlock, (err, newLastBlock) => {
+	let deletedBlockTransactions;
+
+	async.series(
+		{
+			popLastBlock(seriesCb) {
+				// Perform actual delete of last block
+				__private.popLastBlock(lastBlock, (err, previousBlock) => {
 					if (err) {
 						library.logger.error('Error deleting last block', lastBlock);
+					} else {
+						// Store actual lastBlock transactions in reverse order
+						deletedBlockTransactions = lastBlock.transactions.reverse();
+
+						// Set previous block as our new last block
+						lastBlock = modules.blocks.lastBlock.set(previousBlock);
 					}
-					return setImmediate(waterCb, err, newLastBlock);
+					return setImmediate(seriesCb, err);
 				});
 			},
-			(newLastBlock, waterCb) => {
-				modules.system.update(() => setImmediate(waterCb, null, newLastBlock));
+			updateSystemHeaders(seriesCb) {
+				// Update our own headers: broadhash and height
+				modules.system.update(seriesCb);
```

Comment:
I suggest that we should move call to `modules.transport.broadcastHeaders` inside `modules.system.update`, so when ever and where eve we update system headers, we should broadcast it to peers automatically.